Core/Core/Metrics/AbstractMetricsPublisher.cs

```diff
@@ -0,0 +1,89 @@
+﻿using System;
+using System.Json;
+using System.Threading.Tasks;
+using AeroGear.Mobile.Core.Logging;
+using AeroGear.Mobile.Core.Utils;
+
+namespace AeroGear.Mobile.Core.Metrics
+{
+    public abstract class AbstractMetricsPublisher : IMetricsPublisher
+    {
+        private static readonly ILogger LOGGER = MobileCore.Instance.Logger;
+
+        private const string STORAGE_NAME = "org.aerogear.mobile.metrics";
+        private const string STORAGE_KEY = "metrics-sdk-installation-id";
+
+        private readonly IMetrics[] DefaultMetrics;
+        private static readonly DateTime Jan1st1970 = new DateTime(1970, 1, 1, 0, 0, 0, DateTimeKind.Utc);
+
+        public AbstractMetricsPublisher()
+        {
+            DefaultMetrics = new IMetrics[] { new AppMetrics(), new DeviceMetrics() };
+        }
+
+        private static long CurrentTimeMillis()
+        {
+            return (long)(DateTime.UtcNow - Jan1st1970).TotalMilliseconds;
+        }
+
+
+        /// <summary>
+        /// Get or create the client ID that identifies a device as long as the user doesn't reinstall
+        /// the app or delete the app storage. A random UUID is created and stored in the application
+        /// shared preferences.
+        /// </summary>
+        /// <returns>Client ID.</returns>
+        private static String GetClientId()
+        {
+            IUserPreferences preferences = ServiceFinder.Resolve<IPlatformBridge>().GetUserPreferences(STORAGE_NAME);
+
+            String clientId = preferences.GetString(STORAGE_KEY, null);
+
+            if (clientId == null)
+            {
+                clientId = System.Guid.NewGuid().ToString();
+                LOGGER.Info("Generated a new client ID: " + clientId);
+                preferences.PutString(STORAGE_KEY, clientId);
+            }
+
+            return clientId;
+        }
+
+        /**
+         * Parse metrics into a JSONObject and add common information for all metrics requests:
+         *
+         * @param type the type of metrics
+         * @param metrics metrics
+         * @return a JSONObject
+         */
+        protected JsonObject createMetricsJSONObject(string type, IMetrics[] metrics)
+        {
+            JsonObject json = new JsonObject();
+
+            json.Add("clientId", GetClientId());
+            json.Add("timestamp", CurrentTimeMillis());
+            json.Add("type", type);
+
+            JsonObject data = new JsonObject();
+
+            // first put the default metrics (app and device info)
+            foreach (IMetrics m in DefaultMetrics)
+            {
+                data.Add(m.Identifier(), m.Data());
+            }
+
+            // then put the specific ones
+            foreach (IMetrics m in metrics)
+            {
+                data.Add(m.Identifier(), m.Data());
+            }
+
+            json.Add("data", data);
+
+            return json;
+
+        }
+
+        public abstract Task Publish(string type, IMetrics[] metrics);
```

Comment:
I'm not very experienced in C#, so I'm not sure, but shouldn't we use `async` keyword there instead of returning `Task`?
Or this Task is being planned in some different thread executor?

Reply:
Nope, async is to be used when you have to `await` for a task execution (a non async method can't use the `await` keyword).
In this case I return a Task because I want it to be executed in background. If a method needs the result, it must be `async` and will be able to `await` this task

Reply:
ok